relay-server/src/http.rs

```diff
@@ -0,0 +1,299 @@
+///! Abstractions for dealing with either actix-web client structs or reqwest structs.
+///!
+///! All of it is implemented as enums because if they were traits, they'd have to be boxed to be
+///! transferrable between actors. Trait objects in turn do not allow for consuming self, using
+///! generic methods or referencing the Self type in return values, all of which is very useful to
+///! do in builder types.
+///!
+///! Note: This literally does what the `http` crate is supposed to do. That crate has builder
+///! objects and common request objects, it's just that nobody bothers to implement the conversion
+///! logic.
+use std::io;
+use std::io::Write;
+
+use actix_web::client::{ClientRequest, ClientRequestBuilder, ClientResponse};
+use actix_web::http::{ContentEncoding, StatusCode};
+use actix_web::{Binary, Error as ActixError, HttpMessage};
+use brotli2::write::BrotliEncoder;
+use flate2::write::{GzEncoder, ZlibEncoder};
+use flate2::Compression;
+use futures::{future, prelude::*};
+use futures03::{FutureExt, TryFutureExt};
+use serde::de::DeserializeOwned;
+
+use ::actix::prelude::*;
+
+use relay_config::HttpEncoding;
+
+pub enum Request {
+    Actix(ClientRequest),
+    Reqwest(reqwest::Request),
+}
+
+pub enum RequestBuilder {
+    Actix(ClientRequestBuilder),
+    Reqwest {
+        builder: reqwest::RequestBuilder,
+
+        /// The content encoding that this builder object implements on top of reqwest, which does
+        /// not support request encoding at all.
+        http_encoding: HttpEncoding,
+    },
+}
+
+impl RequestBuilder {
+    pub fn reqwest(builder: reqwest::RequestBuilder) -> Self {
+        RequestBuilder::Reqwest {
+            builder,
+
+            // very few endpoints can actually deal with request body content-encoding. Outside of
+            // store/envelope this is almost always identity because there's no way to do content
+            // negotiation on request.
+            http_encoding: HttpEncoding::Identity,
+        }
+    }
+
+    pub fn actix(builder: ClientRequestBuilder) -> Self {
+        RequestBuilder::Actix(builder)
+    }
+
+    pub fn no_default_headers(&mut self) -> &mut Self {
+        match self {
+            RequestBuilder::Actix(builder) => {
+                builder.no_default_headers();
+            }
+            RequestBuilder::Reqwest { .. } => {
+                // can only be set on client
```

Comment:
Since this is not available in Reqwest, it means that some headers would be missing by default. Can we make this a flag (if it even makes senses) that controls these default headers for reqwest, too? At least for the transitionary period.

Reply:
I'll have to investigate why we use this in actix in the first place. As far as I remember actix continued to send its own User-Agent header *in addition* even if explicitly set in the request builder, and that caused all kinds of problems when forwarding encoded requests. I suspect reqwest may be a little bit smarter.

Reply:
I was unable to reproduce any problems with just leaving out the call to `no_default_headers` in all cases, but I'll keep it in for actix and leave it out for reqwest for now

Reply:
IIRC, default headers included compression, content-length, content-type, etc. You can look through the actix code to check which headers it would add and then assess per header.

Reply:
One important header I forgot: `Accept`. That way, Relay could compress its responses to the upstream, which only makes sense in the case of project configs.